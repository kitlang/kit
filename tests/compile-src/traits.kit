trait TestTrait {
    function a(): Void;
    function b(a: Int): Int;
}

implement TestTrait for Int {
    public function a() {
        printf("Int's TestTrait method a\n");
    }

    public function b(x: Int) {
        printf("Int's TestTrait method b\n");
        return this + x;
    }
}


implement TestTrait for CString {
    public function a() {
        printf("CString %s's TestTrait method a\n", this);
    }

    public function b(x: Int) {
        printf("CString %s's TestTrait method b\n", this);
        return x;
    }
}

function main() {
    // static trait dispatch
    (1_i).TestTrait.a();
    printf("%i\n", (1_i).TestTrait.b(2));

    var s: CString = "hello";
    s.TestTrait.a();
    printf("%i\n", s.TestTrait.b(3));

    var x: Int = 777;
    var p: Ptr[Int] = &x;
    p.TestTrait.a();
    printf("%i\n", p.TestTrait.b(5));
}
