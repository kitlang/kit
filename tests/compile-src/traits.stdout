Int's TestTrait method a
Int's TestTrait method b
3
CString hello's TestTrait method a
CString hello's TestTrait method b
3
Int's TestTrait method a
Int's TestTrait method b
782
